### ISSUE TITLE
no-fallthrough: handle upstream changes in control flow analysis (TS 3.7)

### ISSUE THREAD

**Issue by AlCalzone** [fimbullinter/wotan#651]
Repro:

```ts
declare function fail(): never; // throws or something
switch (Math.random()) {
	case 0:
		fail();
		0; // <- detected as unreachable by TS
	case 1: // <- wotan: missing break before case
		break;
	default:
		console.log();
}
```
case 0 will never fall through to case 1

related to #644

**Comment by ajafff** [comment]
Thanks for reporting. I would have forgotten about this rule. This also affects `no-unreachable-code`.

The current implementation is completely based on the AST (as was the old implementation of the control flow graph in TypeScript). The new implementation needs to use type information. That's a breaking change, although there's no way around that.

**Comment by ajafff** [comment]
The rule also needs to handle nested exhaustive switch statements:

```ts
function test(a: unknown, v: 1 | 2): number {
  switch (a) {
    case 1:
      switch (v) {
        case 1:
        case 2:
          return 1; // exhaustive
      }
    case 2: // there's no fallthrough above
      return 2;
  }
}
```

**Comment by ajafff** [comment]
Forget my previous comment. That doesn't seem to work.

**Comment by ajafff** [comment]
:tada: This issue has been resolved in version [v0.23.0](https://github.com/fimbullinter/wotan/releases/tag/v0.23.0) :tada:
